sklearn/ensemble/_hist_gradient_boosting/_binning.pyx

```diff
@@ -70,4 +75,11 @@ cdef void _map_num_col_to_bins(const X_DTYPE_C [:] data,
                     right = middle
                 else:
                     left = middle + 1
-            binned[i] = left
+
+            if is_categorical and binning_thresholds[left] != data[i]:
```

Comment:
If we rely on `_find_binning_threshold` for categorical features as I suggested, this would need to be slightly updated, and we'd need to pass the set of known categories to this function. This would be cleaner and faster since we could directly map unknown categories right from the beginning.